libminifi/src/FlowController.cpp

```diff
@@ -70,60 +71,33 @@ namespace apache {
 namespace nifi {
 namespace minifi {
 
-#define DEFAULT_CONFIG_NAME "conf/config.yml"
-
-FlowController::FlowController(std::shared_ptr<core::Repository> provenance_repo, std::shared_ptr<core::Repository> flow_file_repo, std::shared_ptr<Configure> configure,
-                               std::unique_ptr<core::FlowConfiguration> flow_configuration, std::shared_ptr<core::ContentRepository> content_repo, const std::string name, bool headless_mode)
+FlowController::FlowController(std::shared_ptr<core::Repository> provenance_repo, std::shared_ptr<core::Repository> flow_file_repo,
+                               std::shared_ptr<Configure> configure, std::unique_ptr<core::FlowConfiguration> flow_configuration,
+                               std::shared_ptr<core::ContentRepository> content_repo, const std::string name, bool headless_mode,
+                               std::shared_ptr<utils::file::FileSystem> filesystem)
     : core::controller::ControllerServiceProvider(core::getClassName<FlowController>()),
-      root_(nullptr),
+      c2::C2Client(std::move(configure), std::move(provenance_repo), std::move(flow_file_repo),
+                   std::move(content_repo), std::move(flow_configuration), std::move(filesystem)),
       running_(false),
       updating_(false),
-      c2_enabled_(true),
       initialized_(false),
-      provenance_repo_(provenance_repo),
-      flow_file_repo_(flow_file_repo),
       controller_service_map_(std::make_shared<core::controller::ControllerServiceMap>()),
       thread_pool_(2, false, nullptr, "Flowcontroller threadpool"),
-      flow_configuration_(std::move(flow_configuration)),
-      configuration_(std::move(configure)),
-      content_repo_(content_repo),
       logger_(logging::LoggerFactory<FlowController>::getLogger()) {
-  if (provenance_repo == nullptr)
+  if (IsNullOrEmpty(provenance_repo_))
```

Comment:
the whole existence of `IsNullOrEmpty` is questionable to me, but its usage is ubiquitous, and since it was already used for `configuration_` I rewrote the others as well, I am not against getting rid of it, do you think we should do it, either only here, or in other places as well?

Reply:
`IsNullOrEmpty` could make sense eg. for a `shared_ptr<string>`, but yes, I would remove it in the case where it simply tests whether the argument is null, including `configuration_` -- but not in the is pull request.

In this PR, I would leave `IsNullOrEmpty` calls as they are for now, but wouldn't add new ones.

Reply:
should I remove these newly added `IsNullOrEmpty` calls?

(I agree that, with `string` it kind of made sense to use the empty string to indicate "no-string" (or some other special meaning), now that we have `optional<...>`, I am not so sure about that)

Reply:
Yeah, I would rather remove the `IsNullOrEmpty` calls for `configuration_` and not add any new usages, but I'm OK with keeping it as it is and dealing with this later in another pull request, if you prefer that.

Reply:
rewrote them to use comparison to `nullptr`